### PR TITLE
manual compiler rebase

### PR DESCRIPTION
**Pull request by wsphillips**
I've merged in changes from CUDAnative.jl#master. Please have a look. If you think it's alright I will continue through the source, applying similar merges throughout to incorporate the (many) Julia 1.3+ changes and generic LLVM code. 

In this change I chose not to touch `compiler/device.jl` This is because the diff between CUDAnative and what's there is quite substantial and its harder to tell which sections you built for AMD compatibility and which is just old CUDAnative code that could be updated. If you can flag anything important there with a comment tag in the source, it would be helpful.

### CODE REVIEW

File: src/compiler/irgen.jl [hunk 1]
```diff
@@ -4,17 +4,39 @@ function module_setup(mod::LLVM.Module)
...
     # add debug info metadata
-    push!(metadata(mod), "llvm.module.flags",
-         MDNode([ConstantInt(Int32(1), JuliaContext()),    # llvm::Module::Error
-                 MDString("Debug Info Version"),
-                 ConstantInt(DEBUG_METADATA_VERSION(), JuliaContext())]))
+    if LLVM.version() >= v"8.0"
+    # Set Dwarf Version to 2, the DI printer will downgrade to v2 automatically,
+    # but this is technically correct and the only version supported by NVPTX
+    LLVM.flags(mod)["Dwarf Version", LLVM.API.LLVMModuleFlagBehaviorWarning] =
+        Metadata(ConstantInt(Int32(2), JuliaContext()))
```
Comment (vchuravy) [inline src/compiler/irgen.jl]: We probably don't want to limit DWARF to only 2 here. Haven't checked what AMDGPU supports, but debuginfo support was mimicked on what clang does for nvptx, which might be not right for AMDGPU

**Comment by jpsamaroo** [comment]
Thanks for doing this! I think doing this once, before splitting out reuseable functionality from CUDAnative, would make sense, so that this package can at least be less out of date when we inevitably do so. I'll take a look at `compiler/device.jl` momentarily and let you know what parts that differ need to be kept.

**Comment by maleadt** [comment]
> @maleadt I still feel like extracting the compiler infrastructure out would make sense

For sure, I just don't see it happening right now. Maybe once we have support for multiple packages within a single repo we can try and move out infrastructure piece by piece without too much friction.

**Comment by jpsamaroo** [comment]
@wsphillips would you mind recording the CUDAnative git hash you're working with in this PR somewhere (maybe a short note in the README)?

**Comment by wsphillips** [comment]
> @wsphillips would you mind recording the CUDAnative git hash you're working with in this PR somewhere (maybe a short note in the README)?

Sure. Will do.



> > @maleadt I still feel like extracting the compiler infrastructure out would make sense
> 
> For sure, I just don't see it happening right now. Maybe once we have support for multiple packages within a single repo we can try and move out infrastructure piece by piece without too much friction.

@maleadt Since a lot of what's happening is just whole-sale copy paste of the generic portions of code, do you think you could annotate CUDAnative with a small comment tag in front of code blocks that are stricly Nvidia-related fixes? Sometimes its obvious, sometimes it's not certain (especially for me coming in relatively fresh)

As I am going through the rest of the AMDGPUnative source files, I can do the same. Perhaps it could be a low-effort useful way of visualizing what functionality/filtering/codegen modification is truly universal and what is happening to cater toward vendor-specific needs

File: src/compiler/irgen.jl (hunk 1 above)
Comment (wsphillips) [reply src/compiler/irgen.jl]: Done

File: src/compiler/rtlib.jl [hunk 2]
```diff
@@ -44,7 +44,7 @@ function load_device_libs(agent)
         "hip.amdgcn.bc",
         "irif.amdgcn.bc",
         "ockl.amdgcn.bc",
-        "oclc_isa_version_803.amdgcn.bc", # FIXME: Load based on agent name!
+        "oclc_isa_version_906.amdgcn.bc", # FIXME: Load based on agent name!
```
Comment (wsphillips) [inline src/compiler/rtlib.jl]: Note I modified 803 to 906...change back to your dev when debugging

**Comment by wsphillips** [comment]
> Thanks for doing this! I think doing this once, before splitting out reuseable functionality from CUDAnative, would make sense, so that this package can at least be less out of date when we inevitably do so. I'll take a look at `compiler/device.jl` momentarily and let you know what parts that differ need to be kept.

@jpsamaroo If you're too busy, let me know and I will go ahead and (do my best) to attempt working out the kinks in `compiler/device.jl`

File: src/compiler/rtlib.jl (hunk 2 above)
Comment (jpsamaroo) [reply src/compiler/rtlib.jl]: Yep that's totally fine for now; although if you're feeling adventurous, we should fix this to load whatever agent ISA is in the job struct.

**Comment by jpsamaroo** [comment]
Assuming you meant `compiler/driver.jl` (I don't have a `device.jl` locally), here's a WIP that should get us most of the way there: https://gist.github.com/jpsamaroo/f91f75d5933767daa71a3aa1f49ae308

Note that it has a few "FIXME"s that need addressing before it's ready to go.

**Comment by wsphillips** [comment]
> Assuming you meant `compiler/driver.jl` (I don't have a `device.jl` locally), here's a WIP that should get us most of the way there: https://gist.github.com/jpsamaroo/f91f75d5933767daa71a3aa1f49ae308
> 
> Note that it has a few "FIXME"s that need addressing before it's ready to go.

Yes, that's the file I meant. Sorry--propagation of error :sweat_smile: I merged in your revision. Now the precompile and at least a few tests are passing. I will continue hunting down the remaining issues in the coming days. Thanks!

**Comment by jpsamaroo** [comment]
Awesome! If you're getting some kernels passing (especially the vecadd test), then the rest are probably failing because of the device_libs stuff not being linked in yet (so you'll get errors that a symbol isn't defined or something like that). Note that unlike CUDAnative, we have multiple device libs to link, so we'll need to slightly modify the driver.jl code to take that into account.

**Comment by wsphillips** [comment]
@jpsamaroo everything passing _until_ vadd kernel launches now. Something screwed up in `execution.jl` Have a look?
[stacktrace.txt](https://github.com/JuliaGPU/AMDGPUnative.jl/files/3780616/stacktrace.txt)

**Comment by jpsamaroo** [comment]
The main issue is just how our execution pipeline is calling `compile` (missing the first `target` argument, which should be `:roc`), however I'm running into various issues as I try to get things passing. Let me try to get those sorted out, and then we can share the work of fixing the device libs tests (the math intrinsics tested are the only ones that test device libs right now).

**Comment by wsphillips** [comment]
Migrating over to a local branch from my fork.

**Pull request closed without merging**